libs/browser/src/ipc.rs

```diff
@@ -0,0 +1,263 @@
+#[derive(Debug, Deserialize)]
+#[serde(tag = "type")]
+pub enum Message {
+    /// Log a message to the console.
+    #[serde(rename_all = "camelCase")]
+    Log { message: String },
+
+    /// Quit the application.
+    Quit,
+
+    /// Capture a screenshot of the current web view.
+    #[serde(rename_all = "camelCase")]
+    CaptureScreenshot {
+        #[serde(flatten)]
+        options: CaptureScreenshotOptions,
+    },
+
+    /// Show an open file dialog and return the selected file paths.
+    #[serde(rename_all = "camelCase")]
+    ShowOpenDialog {
+        #[serde(flatten)]
+        options: ShowOpenDialogOptions,
+    },
+}
+
+/// Options for the `kiosk.captureScreenshot` API.
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct CaptureScreenshotOptions {
+    /// The ID of the message to reply to. Internal use only.
+    reply_to: String,
+}
+
+/// Options for the `kiosk.showOpenDialog` API.
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct ShowOpenDialogOptions {
+    /// The ID of the message to reply to. Internal use only.
+    reply_to: String,
+
+    /// The actual options for the file chooser dialog.
+    #[serde(flatten)]
+    inner: ui::FileChooserOptions,
+}
+
+/// Replies that can be sent from the Rust application to the [`WebView`].
+#[derive(Debug, Serialize)]
+#[serde(tag = "type")]
+pub enum Reply {
+    /// Reply with a captured screenshot.
+    #[serde(rename_all = "camelCase")]
+    CaptureScreenshot {
+        /// The ID of the message this is a reply to. Internal use only.
+        in_reply_to: String,
+
+        /// The screenshot data as a PNG image (base64-encoded in JSON).
+        #[serde(with = "Base64Standard")]
+        data: Vec<u8>,
+    },
+
+    /// Reply with the selected file paths from an open file dialog.
+    #[serde(rename_all = "camelCase")]
+    ShowOpenDialog {
+        /// The ID of the message this is a reply to. Internal use only.
+        in_reply_to: String,
+
+        /// Whether the dialog was canceled.
+        canceled: bool,
+
+        /// The selected file paths.
+        file_paths: Vec<String>,
+    },
+
+    /// Reply for errors that occurred during processing.
+    #[serde(rename_all = "camelCase")]
+    Error { message: String },
+}
+
+/// Set up `kiosk.*` APIs in the [`WebView`] and the Rust handler.
+pub fn setup(window: &gtk::Window, webview: &WebView) -> anyhow::Result<()> {
+    let manager = webview
+        .user_content_manager()
+        .context("Failed to get user content manager")?;
+    manager.connect_script_message_received(None, {
+        let window = window.clone();
+        let webview = webview.clone();
+        move |_, message| {
+            if let Some(value) = message.js_value() {
+                gtk::glib::spawn_future_local({
+                    let window = window.clone();
+                    let webview = webview.clone();
```

Comment:
Yeah, it's the `move`s and the fact that the `connect_script_message_received` callback can be called multiple times. Each time it's called, we need to do these inner clones to pass off to the `async` block. The clone is, as you guessed, just copying a pointer around. It may be using something similar to an `Rc` or `Arc` where it's doing reference counting, but I'm not sure.